numpy/testing/utils.py

```diff
@@ -1937,3 +1964,155 @@ def __exit__(self, *exc_info):
                 mod.__warningregistry__.clear()
             if mod in self._warnreg_copies:
                 mod.__warningregistry__.update(self._warnreg_copies[mod])
+
+
+class suppressed_warnings(object):
+    """
+    Context manager doing much the same as ``warnings.catch_warnings``.
+    However, it also provides a suppression mechanism to work around
+    http://bugs.python.org/issue4180.
+
+    This bug causes Python before 3.4 to not realiably show warnings again
+    after they have been ignored once (even within catch_warnings). It
+    means that no "ignore" filter can be used easily, since following
+    tests might need to see the warning.
+
+    Paremeters
+    ----------
+    record : boolean
+        Whether to print or instead record the unsuppressed warnings.
+        If true, the record is available through ``suppressed_warnings().log``
+
+    Notes
+    -----
+    Suppressions added inside the the context manager will be discarted
+    again when leaving it. Upon entering all suppressions defined outside
+    a context will be applied automatically.
+
+    In non-recording mode, unsuppressed warnings are passed to the
+    outer warning handler (i.e. to printing). In recording mode, the
+    `log` attribute will store them.
+
+    If suppressions are added with a module information, the warning
+    registry of this module will additionally be cleared when applying it,
+    entering the context or exiting it.
```

Comment:
The existing warning registry for the module is cleared upon entering? That doesn't sound like wanted behavior I think. But maybe it's a textual issue in this sentence.